File: Crash it/Shuttle/ShuttleEnemiesGenerator.swift
//
//  ShuttleEnemiesGenerator.swift
//  Crash it
//
//  Created by Olivier Picard on 18/03/2018.
//  Copyright © 2018 Olivier Picard. All rights reserved.
//

import Foundation
import SpriteKit

class ShuttleEnemyGenerator {
    /** Nombre d'image de vaisseau comptenu dans l'atlas */
    private let NUMBER_OF_IMAGE_SHUTTLE: UInt32 = 5
    private let scene: SKScene
    private let target: ShuttlePlayer
    
    private var previous_generate_time: TimeInterval?
    private var maxTime_before_generate: TimeInterval = 0.5 // en seconde
    var enemies_percent: Int = 10
    
    
    
    init(scene: SKScene, target: ShuttlePlayer) {
        self.scene = scene
        self.target = target
        arc4random_stir()
    }
    
    func generate(_ currentTime: TimeInterval) {
        if currentTime - (previous_generate_time ?? 0) < maxTime_before_generate { return }
        self.previous_generate_time = currentTime
        if arc4random_uniform(101) > self.enemies_percent { return }
        
        
        let yPos = scene.size.height + 10
        let xPos = CGFloat(arc4random_uniform(UInt32(self.scene.size.width)))
        let image = randomEnemyShuttle(imageIdFrom: 1, imageIdTo: NUMBER_OF_IMAGE_SHUTTLE)
        let shuttle_infos = determineShuttleGlobalInformation(img: image)
        let shuttle = ShuttleEnemy(image: image,
                                   color: shuttle_infos.color,
                                   stats: shuttle_infos.stats,
                                   target: self.target)
        
        shuttle.position = CGPoint(x: xPos, y: yPos)
        self.scene.addChild(shuttle)
    }
    
    
    func randomEnemyShuttle(imageIdFrom: UInt32, imageIdTo: UInt32) -> UIImage {
        let imageName = "ennemy_" +
            String(arc4random_uniform(imageIdTo + 1 - imageIdFrom) + imageIdFrom)
        
        return UIImage(named: imageName)!
    }
    
    
    func determineShuttleGlobalInformation(img: UIImage) -> (stats: Shuttle.Stats, color: UIColor) {
        let stats: Shuttle.Stats
        let color: UIColor
        
        switch img {
        case #imageLiteral(resourceName: "ennemy_1"):
            stats = Shuttle.Stats(defense: 70, attack: 30,
                                  shootStats: Shuttle.Stats.ShootStats(deltaTimeToShoot: 1/5,
                                                                       probSuccessFire: 30))
            color = UIColor.orange
        case #imageLiteral(resourceName: "ennemy_2"):
            stats = Shuttle.Stats(defense: 80, attack: 10,
                                  shootStats: Shuttle.Stats.ShootStats(deltaTimeToShoot: 1/3,
                                                                       probSuccessFire: 40))
            color = UIColor.cyan
        case #imageLiteral(resourceName: "ennemy_3"):
            stats = Shuttle.Stats(defense: 120, attack: 10,
                                  shootStats: Shuttle.Stats.ShootStats(deltaTimeToShoot: 1/3,
                                                                       probSuccessFire: 30))
            color = UIColor.cyan
        case #imageLiteral(resourceName: "ennemy_4"):
            stats = Shuttle.Stats(defense: 230, attack: 10,
                                  shootStats: Shuttle.Stats.ShootStats(deltaTimeToShoot: 1/3,
                                                                       probSuccessFire: 40))
            color = UIColor.cyan
        case #imageLiteral(resourceName: "ennemy_5"):
            stats = Shuttle.Stats(defense: 70, attack: 20,
                                  shootStats: Shuttle.Stats.ShootStats(deltaTimeToShoot: 1/6,
                                                                       probSuccessFire: 20))
            color = UIColor.green
        default:
            stats = Shuttle.Stats(defense: 0, attack: 0, shootStats: Shuttle.Stats.ShootStats())
            color = UIColor.white
        }
        
        return (stats, color)
    }

    
}

File: Android/app/src/main/java/com/olivierpicard/crachit/Shuttle/ShuttleEnemiesGenerator.java
package com.olivierpicard.crachit.Shuttle;

import android.graphics.Color;

import com.olivierpicard.crachit.DataBaseHandler;
import com.olivierpicard.crachit.GameScene;
import com.olivierpicard.crachit.Graphics.GInterval;
import com.olivierpicard.crachit.Graphics.GPoint;
import com.olivierpicard.crachit.Graphics.GVector;
import com.olivierpicard.crachit.IRestaurable;
import com.olivierpicard.crachit.R;
import com.olivierpicard.crachit.Graphics.GTools;

/**
 * Created by olivierpicard on 06/04/2018.
 */

public class ShuttleEnemiesGenerator implements IRestaurable {
    private class ShuttleConf {
        public Stats stats;
        public int color;
        public String imageName;

        public ShuttleConf(Stats stats, int color, String imageName) {
            this.stats = stats;
            this.color = color;
            this.imageName = imageName;
        }
    }


    /** Nombre d'image de vaisseau comptenu dans l'atlas */
    private final int NUMBER_OF_IMAGE_SHUTTLE = 5;
    private final GameScene scene;
    private final ShuttlePlayer target;

    private long previous_generate_time;
    private long maxTime_before_generate = (long)(0.5*1000); // en milli seconde
    public int enemies_percent = 10;
    public boolean enable = false;


    public ShuttleEnemiesGenerator(GameScene scene, ShuttlePlayer target) {
        this.scene = scene;
        this.target = target;
        this.previous_generate_time = 0;
    }


    public void generate(long currentTime) {
        if(!this.enable) return;
        if(currentTime - this.previous_generate_time < this.maxTime_before_generate) return;
        this.previous_generate_time = currentTime;
        if(GInterval.random(0, 100) > this.enemies_percent) return;

        final int yPos = -10;
        final int xPos = GInterval.random(0, (int)this.scene.getSize().width);
        final int image = randomEnemyShuttle(1, NUMBER_OF_IMAGE_SHUTTLE);
        final ShuttleConf shuttle_info = determineShuttleGlobalInformation(image);
        final ShuttleEnemy shuttle = new ShuttleEnemy(image,
                shuttle_info.color, shuttle_info.stats, shuttle_info.imageName, this.target);
        shuttle.setPosition(new GPoint(xPos, yPos));
        this.scene.addChild(shuttle);
    }


    private int randomEnemyShuttle(int imageIdFrom, int imageIdTo) {
        final String imageName = "ennemy_"
                + String.valueOf(GInterval.random(imageIdFrom, imageIdTo));
        return getImageID(imageName);
    }

    public int getImageID(String bitmapName) {
        return GTools.resources.getIdentifier(bitmapName,
                "drawable", "com.olivierpicard.crachit");
    }


    private ShuttleConf determineShuttleGlobalInformation(int imageID) {
        Stats stats = null;
        int color = Color.WHITE;
        String imageName = "";

        switch (imageID) {
            case R.drawable.ennemy_1:
                stats = new Stats(90, 15,
                        new ShootStats((long)(1.0/5*1000), 60));
                color = Color.YELLOW;
                imageName = "ennemy_1";
                break;
            case R.drawable.ennemy_2:
                stats = new Stats(100, 10,
                        new ShootStats((long)(1.0/3*1000), 30));
                color = Color.CYAN;
                imageName = "ennemy_2";
                break;
            case R.drawable.ennemy_3:
                stats = new Stats(120, 10,
                        new ShootStats((long)(1.0/4*1000), 25));
                color = Color.CYAN;
                imageName = "ennemy_3";
                break;
            case R.drawable.ennemy_4:
                stats = new Stats(230, 10,
                        new ShootStats((long)(1.0/3*1000), 25));
                color = Color.CYAN;
                imageName = "ennemy_4";
                break;
            case R.drawable.ennemy_5:
                stats = new Stats(90, 20,
                        new ShootStats((long)(1.0/6*1000), 25));
                color = Color.GREEN;
                imageName = "ennemy_5";
                break;
        }

        return new ShuttleConf(stats, color, imageName);
    }


    @Override
    public void restaure(DataBaseHandler.ItemRestaurationTable item) {
        final ShuttleConf shuttle_info = determineShuttleGlobalInformation(getImageID(item.option1));
        final ShuttleEnemy shuttle = new ShuttleEnemy(getImageID(item.option1),
                shuttle_info.color, shuttle_info.stats, shuttle_info.imageName, this.target);
        shuttle.setPosition(new GPoint(item.xPos, item.yPos));
        shuttle.setZPosition(item.zPosition);
        shuttle.setZRotation(item.zRotation);
        shuttle.direction = new GVector(item.dx, item.dy);
        shuttle.lifeBar.setValue(item.life);
        this.scene.addChild(shuttle);
    }
}
